Intro to git: Content (#40)
Ada-Developers-Academy/core-unit-1: intro-to-git/intro-to-git-vocab-and-concepts.md

# Intro to Git: Vocab and Concepts

## Learning Goals

- Explain how Git is version control system used to version code
- Define a repository in Git
- Compare a Git repository to a project
- Define a commit in Git

## Introduction

> Recall that _versions_ in software are different states of a codebase.

Author: You're right that I was doing a callback to that packaging lesson!

I'll likely add in your suggestion without a link back... Link backs to other resources is not my favorite dependency to manage ;)
(I'm making it sound like it's just my preference; our curriculum suffers when we don't update resources linking to each other, and makes our curriculum less trustworthy to students for good reason)

> Recall that _versions_ in software are different states of a codebase.

Reviewer: Ack! Noticed in the Git notification of your comment that a typo squeaked through on line 5:

Explain how Git is version control system used to version code

should be

Explain how Git is a version control system used to version code

or

Explain how the Git version control system is used to version code

(I like the latter better)

> Consider how the version of you from 10 years ago is different from the version of you from 10 minutes ago (which is different from the you of right now). Are those three versions of you very different from each other? Between the you from 10 years ago and the present you, **what was lost? What was gained? And what was modified?**
> 
> Between the you from 10 minutes ago and the present you, what was lost, gained, or modified? The difference between those two versions of you could be huge, or they could be small, incremental, barely noticeable changes.

Author: That's generally why for line 16 I originally put "between the you from 10 minutes ago and the present you" in hopes that nothing very tragic would have happened during that time, but I definitely still see what you're getting at.

Maybe this?

```suggestion
Consider how the version of you from 10 years ago is different from the version of you from 10 minutes ago (which is different from the you of right now). Are those three versions of you very different from each other?

Comparing the you from 10 minutes ago and the present you, have you gained, dropped, or changed anything? The difference between those two versions of you could be huge, or they could be small, incremental, barely noticeable changes.
```

> Between the you from 10 minutes ago and the present you, what was lost, gained, or modified? The difference between those two versions of you could be huge, or they could be small, incremental, barely noticeable changes.

Reviewer: Great! I love the use of dropped instead of lost.  Sounds much more casual!

What a codebase looked like at one point in time could be a different version of the code right now. We can say that there is a version of a codebase that passed three tests, and a version of the same codebase that passed four tests.

How can the idea of versioning help us build software?

## Vocabulary and Synonyms

> | Vocab | Definition | Synonyms | How to Use in a Sentence |
> | ----- | ---------- | -------- | ------------------------ |
> 
> Version
> Version Control System
> Git
> Repository
> Tracking (in Git)
> Commit

Reviewer 2: ```suggestion
| Vocab | Definition | Synonyms | How to Use in a Sentence |
| ----- | ---------- | -------- | ------------------------ |
| Version | Code at a specific point in time labeled with a number based identifier | draft, copy | "I created a separate version of this file so it wouldn't overwrite the previous version."
| Version Control System | A tool that records changes to a file or set of files over time as versions | VCS | "As a project grows, it's wise to use a VCS to keep track of all its changes and progressions.","Good thing the VCS kept track of my previous commits, otherwise, I would have lost all that work!" |
| Git | A distributed version control system that makes snapshots of files or multiple files in a repository to be later referenced as a version. | - | "The problem with *git* jokes is, everyone has their own version" | 
| Repository | A directory that stores project files | Repo, folder, directory | "The project is in a local repository on my machine." |
Commit | An individual change to file(s), identified with a hash | - | "The recent commit fixed the bug in our project" | 
```

Author: I might push back on your synonyms of "version"; I get what you're saying, but it just so happens that a definition of version that was in the context of releases has been given

Also LOL!!!!! Your Git joke LOLL

## Version Control Systems Let Us Version Our Code

Version Control Systems (VCSs) are tools that allow programmers to create, view, and understand different versions of code.

There are plenty of different Version Control Systems, each with their own set of vocabulary and concepts. **Git** is a version control system that has widespread use and acceptance.

Git gives us the tooling to:

- create different versions of our codebases

> - give names, ids, and information between different versions

Author: Tonally this ends up coming off as mega formal, in my opinion? Idk the "as well as recording additional per-version details" adds a mouthful to an intro section that's aimed towards making it disgustingly approachable by being "snappy"

Maybe this?

```suggestion
- give names and ids to different versions
- record additional details between versions
```

> - give names, ids, and information between different versions

Reviewer: 👍

### !end-callout

### VCS Will Affect Our Developer Workflow

Our developer workflow is the process that starts from a programmer reading the requirements, running the tests, writing code, and submitting the code.

Being able to get all the benefits of a version control system will require us to adapt our developer workflow, and use the version control system constantly while coding!

## Each Repository is a Distinct "Project" Code Base

Git has the concept of a **repository,** or repo. A **repo** is a way to describe a project codebase that is being versioned Git.

Often, a repo will directly correlate to a folder on your machine that contains all of a project's files.

When we say to look at someone else's repo, it likely means that they have their own project and codebase that is being versioned with Git.

## Software is Built as a History

When we compare different versions of code to each other, we see how the code shrinks, grows, and changes between them.

In a sense, version control systems let us understand that software is built over time.

> In order to use Git to compare versions, sometimes we will need to tell Git to _track_ different files in a project. This is Git's way of saying that it needs to watch for changes in certain files. Of course we will want Git to watch for changes; that's how we get all the benefits!

Author: I like the wording in general

In this specific case, I _**might**_ not include the seed for gitignore. You know I love a good planted brain seed. But during introduction material, I think many students would take this hint less as a hint/seed, and more of a, "so when do we need to ignore certain files?" which is not a useful question at this moment/a bit distracting

> In order to use Git to compare versions, sometimes we will need to tell Git to _track_ different files in a project. This is Git's way of saying that it needs to watch for changes in certain files. Of course we will want Git to watch for changes; that's how we get all the benefits!

Reviewer: Understood.  It's a fine line.  If you like the rest of the change, I'd suggest just eliding that sentence, then drop the But from the following

```suggestion
For Git to be able to compare versions, we need to tell it to _track_ different files in a project. **Tracking** is our way of telling Git that it needs to watch for changes in certain files. We can tell Git to track individual files, or whole directories. When we want all the benefits of versioning, we should tell Git to watch for changes!
```

### Commits are a Snapshot of History

**A commit** is a representation of a version of code at one point of time.

We will build commits as we build software; we will create commits, and over time, these commits will represent different versions of code and build a history.

Commits not only represent a version of code, but also they contain a lot of extra information. Every commit can also contain the following information:

- A summary of lines of code that were added, deleted, or changed, compared to the last commit
- A way to refer to the last commit, or its "parent" commit
- A unique ID for the commit, often called the commit hash
- A commit message, which will be a string that describes the commit's contents
- The date and time that the commit was made

> Large projects will be made of hundreds or thousands of commits.

Author: I'd want a vocab word at a lower level than "comprise"

```suggestion
Large projects often have hundreds or thousands of commits.
```

> Large projects will be made of hundreds or thousands of commits.

Reviewer: Nice.

## Teams Collaborate by Merging Histories

> In a work day, we want any number of programmers to be able to write code, contribute commits, and build software.
> 
> Git is a version control system that allows for team collaboration. How?
> 
> Collaboration in this sense means that everyone can contribute changes to code, and look at those changes of code.

Author: ```suggestion
Over the course of a work day, a team of programmers writes code, contributes commits, and builds software. In order to work together effectively, we need to contribute our own changes, as well as review changes made by other team members.

We need to collaborate! But how can we do this while we are each working on our own code?

We can enable team collaboration by using Git! Let's examine how.
```

I'm a big believer that weirdly small, spaced out, even one-line paragraphs at the beginning of a section helps "get the brain engine started," grounds them, and revs up a brain from 0-40, so the rest of the section can go 40-65.

> Collaboration in this sense means that everyone can contribute changes to code, and look at those changes of code.

Reviewer: Stop and breathe, then hit 65!  Woo hoo!

> Every codebase that a team will collaborate on will have a **repository,** or a project folder that contains the files, versioned with Git. (This repository will likely be hosted on github.com)

Author: Hehehehehehe. I see this good extra information. I'm gonna re-read the whole lesson, and obviously I love the extra context, but I might push back on it to consider the amount of new info necessary.

```suggestion
very codebase that a team collaborates on will have a **repository**, or a project folder that contains the files, versioned with Git. This repository is often stored on a hosting site such as github.com.
```

> Every codebase that a team will collaborate on will have a **repository,** or a project folder that contains the files, versioned with Git. (This repository will likely be hosted on github.com)

Reviewer: My intent was to help reinforce that git != github. But I think this is fine.

Every developer who wants to contribute code will need access to the code and files, the repo's commit history, and a way to reference the team's repo. In order to get these three things, every developer will **clone that repo** onto their own machine.

> When a developer clones the repo onto their machine, it will create a project folder on their computer, with all of the files inside of it. Then, each developer will write code, run tests, build features, **get updates from the team**, and **make commits.**

Author: At this moment, I'm not quite excited to bring in the word "pull" -- It's more of a vocabulary jump than I'd like to make during an introduction.

Also, I might keep the bold around the phrases... I really like the direction of keeping it only to vocabulary -- in this case, since we're introducing developer workflow that has no parallel to non-programming processes, I really want to stick in their brain that there are new actions to take.

> Then, each dev who has built commits will send the commits to the team's repository.

Author: ```suggestion
Each dev who makes commits will eventually send the commits to the team's repository.
```

Same as above-- I want this intro to feel more like a storybook that they can memorize, rather than learning words.

> Then, each dev who has built commits will send the commits to the team's repository.

Reviewer: The surrounding steps use future tense, so how about 'eventually will send'?  I had to add a new separate comment for the suggest directive.

Recall that each commit includes this information:

- A summary of lines of code that were added, deleted, or changed, compared to the last commit
- A way to refer to the last commit/its "parent" commit

> Using that information, Git (the tool) will go through a series of merging steps to create a unified history of commits.
> 
> Now, with a unified history of commits, the repo's history will grow. Software development with Git is a cycle of this process.

Author: Hmmmm. I also want to get rid of "(the tool)", but I _do_ want to include an alternative. When programmers say "Git does this," "Git does that," a newbie doesn't really get what Git _is_, that it can _do_ things, and end up personifying it as a weird person with its own personality in their head.

I want to get rid of personification/mystification of things like Git, because then people add personality traits to it, "Git is finicky" or "Git doesn't like me." -- I'd like a generation of developers that go "Git is a tool, and if it's not doing what I want it to do, it's because I'm not using the tool correctly"

Maybe to summarize:
Git is a _tool_, and I like emphasizing that often times, tools run whole processes... not because it has a personality, but because that's how the tool is intended to be used.

```suggestion
Using that information, Git must go through a series of merging steps to create a unified history of commits. As each commit is merged in, the repo's history grows. Software development with Git is a cycle of this process.
```

- Changed the verb to "must" to imply a lack of will
- Made it one paragraph -- "Software development with Git is a cycle of this process." on its own line has more of a "mic drop" feel to it than I want lol, and in this case, I'm hoping that "a cycle of this process" is tied to the idea of growing a unified history.

> Now, with a unified history of commits, the repo's history will grow. Software development with Git is a cycle of this process.

Author: sorry for a weirdly poetic response, hopefully it was amusing!

> Now, with a unified history of commits, the repo's history will grow. Software development with Git is a cycle of this process.

Reviewer: Must is a great word choice there!  I confess I don't have the same reaction to the single line "mic drop" 😆, but rolling it into a single paragraphs still reads fine.

I think the goal of avoiding tool personification is going to be a tall order.  I don't know of a single developer who doesn't imbue their tech stack with its own sense of agency.  We know they're all written by other _people_, but by giving them a personality, it gives us room to explain the parts we don't yet understand.  I find git particularly chatty in how it often recommends other commands you might try, and yet very opaque at the same time.  It sometimes requires an act of will to remind myself that the software is just software.  But I think that's something that will need to be reinforced with students repeatedly over time.

In this situation, though, I think the additional changes really well.

### !callout-info

## Comparing Different Ways to Collaborate

> Different versioning tools have different strategies. An older VCS used a system of _checking out a file for a period of time,_ much like checking out a book from a library. This VCS said that other team members couldn't work on, or even **_open_** that file until that dev checked it back in!

Author: Hmmmmm. I'm not sure, I don't have a good opinion on this

### !end-callout

> 
> ## Check for Understanding

Reviewer 2: ```suggestion

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge

* type: multiple-choice
* id: 6ceb3b61-a46f-41a4-b649-5b7a1826b632
* title: Git Vocab I 

##### !question
What is Git? 
##### !end-question

##### !options
* Version Control System
* Distributed Version Control System
* Repository Control System
##### !end-options

##### !answer
* Distributed Version Control System
##### !end-answer

### !end-challenge

<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge

* type: multiple-choice
* id: fb75bb2b-258c-403c-a3ba-6f5923777cbe
* title: Git Vocab II

##### !question
What is an example of a repository?
##### !end-question

##### !options
* a folder with 1 or more files
* a single file
* a file version
##### !end-options

##### !answer
* a folder with 1 or more files
##### !end-answer

### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge

* type: multiple-choice
* id: b0961893-7591-4c5a-9430-0b0c4c0958ae
* title: Git vocab III

##### !question
What is a commit?
##### !end-question

##### !options
* Clones of file(s)
* Version(s) of file(s)
* Folder with multiple files
##### !end-options

##### !answer
* Version(s) of file(s)
##### !end-answer

### !end-challenge
<!-- prettier-ignore-end -->

```

## Teams Collaborate by Merging Histories

Over the course of a work day, a team of programmers writes code, contributes commits, and builds software. In order to work together effectively, we need to contribute our own changes, as well as review changes made by other team members.

We need to collaborate! But how can we do this while we are each working on our own code?

We can enable team collaboration by using Git! Let's examine how.

Every codebase that a team will collaborate on will have a **repository,** or a project folder that contains the files, versioned with Git. (This repository will likely be hosted on github.com)

Every developer who wants to contribute code will need access to the code and files, the repo's commit history, and a way to reference the team's repo. In order to get these three things, every developer will **clone** that repo onto their own machine.

When a developer clones the repo onto their machine, it will create a project folder on their computer, with all of the files inside of it. Then, each developer will write code, run tests, build features, **get updates from the team**, and **make commits.**

> Each dev who makes commits eventually sends the commits to the team's repository.

Reviewer: How about leaving the sends as will send.  The prior steps all use future rather than present tense.
```suggestion
Each dev who makes commits eventually will send the commits to the team's repository.
```
